File: Week_02/G20190343020094/589.n叉树的前序遍历.py
#
# @lc app=leetcode.cn id=589 lang=python3
#
# [589] N叉树的前序遍历
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
# class Solution:
#     def preorder(self, root: 'Node') -> List[int]:
#         res = []
#         if root == None: return res

#         def recursion(root, res):
#             res.append(root.val)
#             for child in root.children:
#                 recursion(child, res)
#         recursion(root, res)
#         return res

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        res = []
        if root == None: return res

        stack = [root]
        while stack:
            curr = stack.pop()
            res.append(curr.val)
            stack.extend(curr.children[::-1])
        return res


        
# @lc code=end

File: Week_02/G20190343020228/589.n叉树的前序遍历.cs
/*
 * @lc app=leetcode.cn id=589 lang=csharp
 *
 * [589] N叉树的前序遍历
 */

// @lc code=start
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val,IList<Node> _children) {
        val = _val;
        children = _children;
    }
}
*/
public class Solution {
    public IList<int> Preorder(Node root){
        List<int> result = new List<int>();
        if(root==null) 
            return result;
        Stack<Node> stack = new Stack<Node>();
        stack.Push(root);
        while(stack.Count>0){
            Node curr = stack.Pop();
            result.Add(curr.val);
            IList<Node> clist = curr.children;
            for(int i = clist.Count-1;i>=0;i--){
                stack.Push(clist[i]);
            }
        }
        return result;
    }
    public IList<int> PreorderWithRecursive(Node root) {
        List<int> result = new List<int>();
        if(root==null) 
            return result;

        helper(root, result);
        return result;
    }

    private void helper(Node node, List<int> result){
        if(node==null) return;

        result.Add(node.val);
        if(node.children!=null&&node.children.Count>0){
            foreach(Node cnode in node.children){
                if(cnode!=null)
                    helper(cnode,result);
            }
        }
    }
}
// @lc code=end

